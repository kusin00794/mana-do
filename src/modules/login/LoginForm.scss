@import '../../styles/variables.scss';

.login-form {
  display: flex;
  width: 500px;
  align-items: center;
  justify-content: center;

  form {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    max-width: 100%;
  }

  input {
    width: 100%;
    padding: 12px 20px;
    margin: 8px 0 24px;
    display: inline-block;
    border: 1px solid $input-border-color;
    border-radius: 4px;
    box-sizing: border-box;
  }

  button {
    width: 100%;
    background: $active-color;
    color: $white-color;
  }
}
