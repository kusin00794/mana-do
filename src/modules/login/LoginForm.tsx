import React, { useState, useEffect } from 'react';
import { useHistory } from 'react-router-dom';

import Service from '../../service';
import RoutesString from '../../pages/routesString';

import './LoginForm.scss';

const LoginForm = () => {
  // move to authentication state
  const isLoggedIn = localStorage.getItem('auth');
  const [form, setForm] = useState({
    userId: '',
    password: '',
  });
  const history = useHistory();

  const signIn = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { userId, password } = form;
    const resp = await Service.signIn(userId, password);

    localStorage.setItem('auth', JSON.stringify({ token: resp, userId }));
    history.push(RoutesString.Welcome);
  };

  const onChangeField = (e: React.ChangeEvent<HTMLInputElement>) => {
    e.persist();
    setForm((prev) => ({
      ...prev,
      [e.target.name]: e.target.value,
    }));
  };

  useEffect(() => {
    if (isLoggedIn) {
      return history.push(RoutesString.Welcome);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div className="login-form">
      <form onSubmit={signIn} className="row">
        <div className="col-12">
          <label htmlFor="user_id">
            User id
            <input id="user_id" name="userId" value={form.userId} onChange={onChangeField} />
          </label>
        </div>
        <div className="col-12">
          <label htmlFor="password">
            Password
            <input id="password" name="password" type="password" value={form.password} onChange={onChangeField} />
          </label>
        </div>
        <div className="col-12">
          <button type="submit">Sign in</button>
        </div>
      </form>
    </div>
  );
};

export default LoginForm;
