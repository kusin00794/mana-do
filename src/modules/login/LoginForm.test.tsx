import React from 'react';
import { render, wait, fireEvent } from '@testing-library/react';

import Service from '../../service';
import { testSnapshots } from '../../utils/test';
import RoutesString from '../../pages/routesString';
import LoginForm from './LoginForm';

const mockToken = 'testabc.xyz.ahk';
const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
  useLocation: () => ({}),
}));

describe('<LoginForm />', () => {
  testSnapshots(LoginForm, [
    {
      props: {},
      description: 'default render',
    },
  ]);
});

describe('Handle Logic', () => {
  test('Should login successfully and redirect user to home page', async () => {
    Service.signIn = jest.fn().mockResolvedValue(mockToken);
    const event = {
      persist: jest.fn(),
    };
    const userInfo = {
      id: 'firstUser',
      password: 'example',
    };
    const wrapper = render(<LoginForm />);
    const { container } = wrapper;
    const userId = container.querySelector('#user_id');
    const userPassword = container.querySelector('#password');
    const button = container.querySelector('button');

    await wait(() => {
      userId &&
        fireEvent.change(userId, {
          ...event,
          target: {
            name: 'userId',
            value: userInfo.id,
          },
        });
      userPassword &&
        fireEvent.change(userPassword, {
          ...event,
          target: {
            name: 'password',
            value: userInfo.password,
          },
        });
      button && fireEvent.click(button);
    });

    expect(mockHistoryPush).toHaveBeenCalledWith(RoutesString.Welcome);
  });
});
