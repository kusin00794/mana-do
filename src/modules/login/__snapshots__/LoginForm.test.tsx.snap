// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> snapshots default render 1`] = `
<div
  className="login-form"
>
  <form
    className="row"
    onSubmit={[Function]}
  >
    <div
      className="col-12"
    >
      <label
        htmlFor="user_id"
      >
        User id
        <input
          id="user_id"
          name="userId"
          onChange={[Function]}
          value=""
        />
      </label>
    </div>
    <div
      className="col-12"
    >
      <label
        htmlFor="password"
      >
        Password
        <input
          id="password"
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        />
      </label>
    </div>
    <div
      className="col-12"
    >
      <button
        type="submit"
      >
        Sign in
      </button>
    </div>
  </form>
</div>
`;
